egs/librispeech/ASR/transducer_stateless/decoder.py

```diff
@@ -0,0 +1,65 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang)
+#
+# See ../../../../LICENSE for clarification regarding multiple authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn as nn
+
+
+class Decoder(nn.Module):
```

Comment:
After analyzing the decoded results, everything seems to be OK except for the following two cases.

## Case 1: Words containing repeated tokens

Example: `odd`, `summon`, `accuse`, `torrent`, `accuracy`, `saddler`, `occupied`, `accent`, `coffee`, and so on.

When `odd` is converted into pieces, it is `_o d d`.

## Case 2:  A word itself is a token and it is also the prefix of the next word

Example: `the thermometer`, which is `_the _the r mo me ter` when breaking into pieces.

In both cases, it gets stuck during decoding as the decoder always predicts the same token.
Take the word `odd` as an example. When we predict `o`, it may update the decoder to predict `d`. Then interesting things happen. If feeding `d` to the decoder, it will predict the next token, which is also `d`. As it is a bigram and does not take further history into account, the decoder goes into an infinite loop.

---

Some of the decoding results using `--epoch 5 --avg 1` (trained with 100 hours subset) is

![Screen Shot 2021-12-18 at 00 25 46](https://user-images.githubusercontent.com/5284924/146576348-416ede9e-e672-4b74-aec8-070c82178cf9.png)


![Screen Shot 2021-12-18 at 00 26 02](https://user-images.githubusercontent.com/5284924/146576400-ddfb5903-a20f-451e-b6b2-f10c4ff5d627.png)

Reply:
I think I'm missing something; why are you saying that the decoder is a bigram? Isn't it an LSTM? It should preserve enough history in the "memory".

Reply:
> Isn't it an LSTM?

Yes, you're right. In a typical RNN-T model, the decoder is an RNN.

There are papers saying the decoder does not function as an LM, e.g., 
- Hybrid Autoregressive Transducer (HAT) ,https://arxiv.org/pdf/2003.07705.pdf
> According to our analysis, the decoder network
does not behave as a language model but more like a finite
context model

- Rnn-Transducer with Stateless Prediction Network (https://ieeexplore.ieee.org/document/9054419)


We are trying to replace the decoder with a feed-forward network and the current file implements the decoder network from the second paper. For easy reference, the following screenshot from the paper shows its architecture:
![Screen Shot 2021-12-18 at 07 30 17](https://user-images.githubusercontent.com/5284924/146619267-3ad61296-e4d1-4dce-a146-2bff4a86e460.png)

You can see from (b) that the decoder contains just an embedding layer.

Reply:
Hm.
I think in that paper they may not have been doing just greedy search, they were actually decoding with some kind of LM.
I suspect this problem might be (mostly) resolvable by using 2 tokens of history instead of 1, e.g. using a conv1d layer with kernel size 2, and manually pad on the left only.

Reply:
> I suspect this problem might be (mostly) resolvable by using 2 tokens of history instead of 1, e.g. using a conv1d layer with kernel size 2, and manually pad on the left only.

Yes, I agree.

I just merged a PR (https://github.com/k2-fsa/icefall/pull/151), which limits the number of symbols per frame, without changing the model architecture. 

---

Will add a convolution layer to the decoder.